### PR TITLE
test: add a bom test for derived images

### PR DESCRIPTION
**Pull request by rchincha**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/bom.bats [hunk 1]
```diff
@@ -106,19 +106,47 @@ bom-parent:
...
     [ -f .stacker/artifacts/bom-parent/installed-packages.json ]
     # a full inventory for this image
     [ -f .stacker/artifacts/bom-parent/inventory.json ]
     # sbom for this image
     [ -f .stacker/artifacts/bom-parent/bom-parent.json ]
+    # a full inventory for this image
+    [ -f .stacker/artifacts/bom-child/inventory.json ]
+    # sbom for this image
+    [ -f .stacker/artifacts/bom-child/bom-child.json ]
     if [ -nz "${REGISTRY_URL}" ]; then
```
Comment (smoser) [inline test/bom.bats:140]: this is bad shell.
i think probably just a typo, and that you meant
```bash
if [ -n "${REGISTRY_URL}" ]; then
fi
```

**Review by smoser — CHANGES_REQUESTED** [review]

File: test/bom.bats [hunk 2]
```diff
@@ -106,19 +106,47 @@ bom-parent:
...
       stacker publish --skip-tls --url docker://localhost:8080/ --tag latest
       refs=$(regctl artifact tree localhost:8080/bom-parent:latest --format "{{json .}}" | jq '.referrer | length')
```
Comment (smoser) [inline test/bom.bats:142]: shouldn't you use REGISTRY_URL here rather than localhost?

File: test/bom.bats [hunk 3]
```diff
@@ -106,18 +106,46 @@ bom-parent:
...
-      stacker publish --skip-tls --url docker://localhost:8080/ --tag latest
-      refs=$(regctl artifact tree localhost:8080/bom-parent:latest --format "{{json .}}" | jq '.referrer | length')
+    # a full inventory for this image
+    [ -f .stacker/artifacts/bom-child/inventory.json ]
+    # sbom for this image
+    [ -f .stacker/artifacts/bom-child/bom-child.json ]
+    if [ -n "${REGISTRY_URL}" ]; then
+      stacker publish --skip-tls --url docker://${REGISTRY_URL}/ --tag latest
+      refs=$(regctl artifact tree ${REGISTRY_URL}/bom-parent:latest --format "{{json .}}" | jq '.referrer | length')
+      [ $refs eq 2 ]
```
Comment (smoser) [inline test/bom.bats:143]: you want `[ $refs -eq 2 ]`

```bash
$ [ 2 eq 2 ]
bash: [: eq: binary operator expected
```

**Review by smoser — CHANGES_REQUESTED** [review]

**Comment by rchincha** [comment]
publishing bom-good.yaml bom-parent to docker://localhost:8080/bom-parent:latest

Copying blob a1d0c7532777 skipped: already exists
Copying blob c80fd6526a2d skipped: already exists
Copying config 3b49bea002 done
Writing manifest to image destination
Storing signatures
artifact '/tmp/stacker/.stacker/artifacts/bom-parent/bom-parent.json' sucessfully uploaded to url:http://localhost:8080/v2/bom-parent/manifests/sha256:99d732817fdb2c707ff140b399bbc7981880618b4dea647d759921b069517057
artifact '/tmp/stacker/.stacker/artifacts/bom-parent/inventory.json' sucessfully uploaded to url:http://localhost:8080/v2/bom-parent/manifests/sha256:a6334f2ed74517fdc7bef5a04ffa1d31f86782547bc5551eed00a63b5610d8fa

I am able to test this locally. However, in our CI ..

2023-08-23T18:48:15.9910388Z # usernsexec-ing [u 0 1001 1 1 165537 65535 g 0 123 1 1 165537 65535 -- /home/runner/work/stacker/stacker/stacker --internal-userns --debug publish --skip-tls --url docker://localhost:5000 --tag latest]
2023-08-23T18:48:15.9910592Z # stacker version 5b6c4e1c692ab2a65663c40380583d6750746337
2023-08-23T18:48:15.9910779Z # initializing stacker recipe: stacker.yaml
2023-08-23T18:48:15.9911307Z # substituting $STACKER_ROOTFS_DIR to /home/runner/work/stacker/stacker/stackertest-test_bom_tool_should_work_inside_run.nexivr/roots
2023-08-23T18:48:15.9911849Z # substituting $STACKER_STACKER_DIR to /home/runner/work/stacker/stacker/stackertest-test_bom_tool_should_work_inside_run.nexivr/.stacker
2023-08-23T18:48:15.9912357Z # substituting $STACKER_OCI_DIR to /home/runner/work/stacker/stacker/stackertest-test_bom_tool_should_work_inside_run.nexivr/oci
2023-08-23T18:48:15.9912517Z # substituting $STACKER_WORK_DIR to 
2023-08-23T18:48:15.9912903Z # publishing stacker.yaml bom-parent to docker://localhost:5000/bom-parent:latest
2023-08-23T18:48:15.9912995Z # 
2023-08-23T18:48:15.9913440Z # subject:http://localhost:5000/v2/bom-parent/manifests/latest doesn't exist, ignoring and proceeding
2023-08-23T18:48:15.9913880Z # unable to get descriptor details for subject:http://localhost:5000/v2/bom-parent/manifests/latest

**Comment by rchincha** [comment]
Maybe zot is not up

File: pkg/stacker/publisher.go [hunk 4]
```diff
@@ -392,24 +450,25 @@ func (p *Publisher) Publish(file string) error {
 
 					registry := url.Host
 					repo := url.Path
+					/*
```
Comment (hallyn) [inline pkg/stacker/publisher.go:453]: Please just cut this code.

File: pkg/stacker/publisher.go [hunk 5]
```diff
@@ -261,7 +234,92 @@ func (p *Publisher) publishArtifact(path, mtype, registry, repo, subject string,
...
+	} else {
+		url = fmt.Sprintf("https://%s/v2%s/manifests/%s", registry, strings.Split(repo, ":")[0], tag)
+	}
+
+	return url
+}
+
+func uploadBlob(registry, repo, path, username, password string, reader io.Reader, size int64, dgst *godigest.Digest, skipTLS bool) error {
+	// upload with POST, PUT sequence
+	var regUrl string
```
Comment (hallyn) [inline pkg/stacker/publisher.go:256]: Not asking you to do that here for this pr, but just a comment that tracking the regUrl and subject as instances of a new custom url class, which knows whether it's skipTLS or not and tracks repo etc, and has .BlobURL and .ManiestURL both returning a string url, would be neat here.

Comment (rchincha) [reply pkg/stacker/publisher.go:256]: Would be neat I suppose.

File: pkg/stacker/publisher.go (hunk 4 above)
Comment (rchincha) [reply pkg/stacker/publisher.go]: Fixed.

**Review by hallyn — APPROVED** [review]

**Pull request merged**